### DISCUSSION TITLE
How to define complex page loading.

### DISCUSSION THREAD

**Discussion by kimFlutter** [rrousselGit/riverpod#3281] [Q&A]
With FutureProvider, it is very easy to use AsyncValue to display loading ui for one FutureProvider.
However, in my app, I am receiving data from home page using 3 api and when I click on a button to change my profile, only 1 out of 3 api is called.
I want to implement the same page circle loading for both of the above situations.

To put it simply, I think you can use stateProvider or use local variables, but I think there will be a smarter way.

What should I do?

To summarize, the requirements are as follows.
1. When entering the page: Display the status of full screen loading on the page until you receive the data from the three apps and receive all the data
2. When switching profiles using buttons: Call 1 of 3 api to receive data and display full screen loading status

**Comment by noriHanda** [comment]
Let's say those three api calls are called in build methods of three AsyncNotifiers, A, B, and C. (And this is the best way to implement IMO, assuming that you need to define side effects on each AsyncNotifiers)
Define a FutureProvider that watches all A, B, and C, and make your UI watch this FutureProvider.
Call methods defined in A, B, and C to perform some side effects.

**Reply by kimFlutter** [reply]
That's a good idea, so it's in this form of code.

```
final someProvider = FutureProvider.autoDispose<void>((ref) async {
  await ref.read(some1Provider.future);
  await ref.read(some2Provider.future);
   await ref.read(some3Provider.future);
  return;
});
```
Can you tell me if I'm trying to come up with the code properly?

**Reply by noriHanda** [reply]
You need to watch those providers instead of read

**Reply by kimFlutter** [reply]
If I do watch instead of read, if any of the providers combined with someProvider are refreshed, won't someProvider run again?
(The use of watch() in official documents is recommended, not compulsory)

**Reply by noriHanda** [reply]
No I don't think so. Your UI won't show a spinner if you read instead of watch.

**Reply by kimFlutter** [reply]
Well... I don't think so. When I actually do it, if I write the syntax with await read(someProvider.future), I can use watch() to receive the data in one-time without observing any unnecessary value changes.
What do you think?

**Reply by noriHanda** [reply]
It shows you a spinner just for the first time with ref.read inside your FutureProvider, but then after that, it won't because your FutureProvider is not watching anything. I thought you wanted to show a spinner every time subscribed providers are in AsyncLoading.
Also, making ref.read ref.watch is almost no-brainer since Riverpod's main philosophy is "reactiveness" as stated in the landing page of the official doc. Following this convention makes your life easier :)

**Reply by noriHanda** [reply]
@kimFlutter You can mark this discussion answered if mine did

**Comment by XuanTung95** [comment]
Have you found one better than StateProvider?
